home/.emacs.d/config.org
#+TITLE:Config
#+AUTH
#+STARTUP: showeverything

* :toc:
- [[#auto-install-packages][auto install packages]]
- [[#common][common]]
- [[#main-packages][main packages]]
  - [[#evil-mode][Evil Mode]]
  - [[#general-keybindings][General keybindings]]
  - [[#pojectile][pojectile]]
  - [[#ivy][Ivy]]
  - [[#key-chord][key chord]]
  - [[#flycheck][flycheck]]
  - [[#dired][dired]]
  - [[#image-dired][image dired]]
  - [[#org-download][org-download]]
  - [[#restclient][restclient]]
  - [[#persp-mode][persp-mode]]
  - [[#anzu][anzu]]
  - [[#doom-modeline][doom-modeline]]
  - [[#perfect-margin][Perfect margin]]
  - [[#centered-cursor-mode][centered-cursor-mode]]
- [[#keybinds][KeyBinds]]
  - [[#file][file]]
  - [[#resize-window][resize window]]
  - [[#marks][marks]]
  - [[#buffer--bookmarks--harpoon][buffer & bookmarks & harpoon]]
  - [[#ivy-1][ivy]]
  - [[#project][project]]
  - [[#dired-1][DIRED]]
  - [[#sudo-edit--code][Sudo Edit & code]]
  - [[#org-mode][ORG mode]]
  - [[#restclient-1][Restclient]]
  - [[#magit][Magit]]
- [[#org-settings][ORG settings]]
  - [[#org-temo-chet-sheet][Org Temo chet sheet]]
  - [[#toc][toc]]
- [[#programming][PROGRAMMING]]
  - [[#python][python]]
  - [[#tree-sitter][tree-sitter]]
- [[#decorations][decorations]]
  - [[#colorthemes][colorthemes]]
  - [[#diff][diff]]
  - [[#fonts][fonts]]
  - [[#icons][icons]]
  - [[#rainbow-mode][rainbow mode]]
  - [[#nyan-mode][Nyan-mode]]
  - [[#evil-snipe][evil-snipe]]
  - [[#org-chckbox-hl][org chckbox hl]]
  - [[#hl-todo][hl-todo]]

* auto install packages
#+begin_src emacs-lisp
(require 'package)
  (add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)

(package-initialize)

(defvar my-required-packages
  '(doom-themes
    evil
    evil-collection
    general
    key-chord
    toc-org
    org-bullets
    projectile
    sudo-edit
    ivy
    counsel
    harpoon
    bm
    vterm
    org-download
    restclient
    persp-mode ;; watch docs
    magit
    magit-delta
    image-dired

    ;; decorations
    rainbow-mode
    all-the-icons
    all-the-icons-dired ;; M-x all-the-icons-install-fonts
    centered-window
    nyan-mode
    perfect-margin
    hl-todo

    tree-sitter
    diff-hl
    dired-subtree
    tree-sitter-langs
    flycheck
    flycheck-pycheckers
    ;; python
    elpy

    haskell-mode
   ))

  ;; Список пакетов для установки.

;; Установите недостающие пакеты.
(dolist (p my-required-packages)
  (unless (package-installed-p p)
    (package-install p)))
#+end_src


* common
#+begin_src emacs-lisp
(menu-bar-mode -1)                                        ;  Отключение меню-бара
(tool-bar-mode -1)                                        ;  Отключение панели инструментов
(scroll-bar-mode -1)                                      ;  Отключение вертикальной полосы прокрутки
(show-paren-mode 1)                                       ;  Включение подсветки соответствующих скобок

(setq make-backup-files nil)                              ;  Отключение создания файлов резервных копий (файлов с суффиксом ~)
(setq auto-save-default nil)                              ;  Отключение автосохранения и создания файлов с префиксом .#
(setq inhibit-startup-screen t)                           ;  Отключение стартового экрана при запуске Emacs

(setq-default isearch-lazy-highlight t)                   ;  Включение "ленивой" подсветки при поиске

(global-visual-line-mode t)                               ;  Включение переноса строк вместо горизонтального скролла
(setq-default display-line-numbers 'relative)             ;  Включение относительных номеров строк

(prefer-coding-system 'utf-8)                             ;  Установка предпочтительной системы кодирования на UTF-8
(set-default-coding-systems 'utf-8)                       ;  Установка системы кодирования по умолчанию на UTF-8
(set-terminal-coding-system 'utf-8)                       ;  Установка системы кодирования терминала на UTF-8
(set-keyboard-coding-system 'utf-8)                       ;  Установка системы кодирования клавиатуры на UTF-8
(setq default-buffer-file-coding-system 'utf-8)           ;  Установка системы кодирования файловых буферов на UTF-8
(setq scroll-conservatively 10000)                        ;  Установка плавного скроллинга без резкого перемещения экрана
(electric-pair-mode 1)                                    ;  автоматически закрывать скопки
(global-hl-line-mode 1)                                   ;  cursor line
(add-hook 'before-save-hook 'delete-trailing-whitespace)  ;  Удаление пустых пробелов в конце строк перед сохранением файла
(blink-cursor-mode 0)                                     ;  cursorline
#+end_src

* main packages
** Evil Mode
https://github.com/emacs-evil/evil
#+begin_src emacs-lisp
;; evil, evil-collection
(setq evil-want-integration t)
(setq evil-want-keybinding nil)
(setq evil-vsplit-window-right t)
(setq evil-split-window-below t)
(setq evil-want-C-u-scroll t)

(require 'evil)
(evil-mode 1)
(setq-default evil-surround-pairs-alist '((?\( . ("(" . ")"))   ;; убираем пробелы
                                          (?\[ . ("[" . "]"))
                                          (?\{ . ("{" . "}"))
                                          (?\" . ("\"" . "\""))
                                          (?\` . ("`" . "`"))
                                          (?\' . ("'" . "'"))
                                          (?\) . ("(" . ")"))
                                          (?\] . ("[" . "]"))
                                          (?\} . ("{" . "}"))))

(define-key evil-normal-state-map "p" 'evil-paste-after) ;  не копировать после вставки
(setq evil-kill-on-visual-paste nil)                     ;  не копировать после вставки
;; (setq-default evil-kill-on-visual-paste nil)
(setq evil-collection-mode-list '(dashboard dired ibuffer))
(evil-collection-init)

(require 'evil-surround)
(global-evil-surround-mode 1)
#+end_src


** General keybindings
https://github.com/noctuid/general.el

#+begin_src emacs-lisp
(require 'general)
(general-evil-setup)

(general-create-definer spc-leader
  :states '(normal insert visual emacs)
  :keymaps 'override
  :prefix "SPC" ;; set leader
  :global-prefix "M-SPC") ;; access leader in insert mode
#+end_src

** pojectile
#+begin_src emacs-lisp
(projectile-mode +1)
(setq projectile-completion-system 'ivy)
#+end_src

** Ivy
#+begin_src emacs-lisp
(ivy-mode +1)
#+end_src

** key chord
#+begin_src emacs-lisp
(require 'key-chord)
(key-chord-mode 1)
(key-chord-define evil-insert-state-map  "jk" 'evil-normal-state)
(key-chord-define evil-insert-state-map  "kj" 'evil-normal-state)
#+end_src

** flycheck
#+begin_src emacs-lisp
(global-flycheck-mode 1)
(with-eval-after-load 'flycheck
  (add-hook 'flycheck-mode-hook #'flycheck-pycheckers-setup))
(flycheck-add-next-checker 'python-flake8 'python-pylint)

(flymake-mode -1)
#+end_src

** dired
#+begin_src emacs-lisp
(require 'dired)
(setq dired-sort-default-list '(("name" "Name" t (t)))
                           dired-listing-switches "-alh --group-directories-first")

(defun dired-open-file-in-new-tab ()
  "Open the current file in dired in a new tab."
  (interactive)
  (tab-new)
  (dired-find-file))

(setq dired-guess-shell-alist-user
      '(("\\.xlsx\\'" "libreoffice")))
#+end_src

** image dired
#+begin_src emacs-lisp
(require 'image-dired)
(setq image-dired-thumb-size 300)
#+end_src

** org-download
#+begin_src emacs-lisp
(require 'org-download)
(setq-default org-download-image-dir "~/Pictures/foo")
(add-hook 'dired-mode-hook 'org-download-enable)
#+end_src

** restclient
#+begin_src emacs-lisp
(require 'restclient)
(add-to-list 'auto-mode-alist '("\\.http\\'" . restclient-mode))
#+end_src

** persp-mode
#+begin_src emacs-lisp
(use-package persp-mode
  :ensure t
  :config
  (setq persp-nil-name "1")
  (with-eval-after-load "persp-mode-autoloads"
    (setq wg-morph-on nil) ;; switch off animation
    ;; (setq persp-autokill-buffer-on-remove 'kill-weak)
    (add-hook 'window-setup-hook #'(lambda () (persp-mode 1))))
  )

(spc-leader
  "w" '(:ignore t :wk "workspace")
  "w c" '(persp-add-new :wk "create workspace")
  "w s" '(persp-switch :wk "switch workspace")
  "w n" '(persp-next :wk "next workspace")
  "w p" '(persp-prev :wk "prev workspace")
  "w k" '(persp-kill :wk "kill workspace")

  "w 1" '((lambda () (interactive) (persp-frame-switch "1")) :wk "1 (main)")
  "w 2" '((lambda () (interactive) (persp-frame-switch "2")) :wk "2 (http)")
  "w 3" '((lambda () (interactive) (persp-frame-switch "3")) :wk "3 (research)")
  "w 4" '((lambda () (interactive) (persp-frame-switch "4")) :wk "4 (other project)")
  "w 5" '((lambda () (interactive) (persp-frame-switch "5")) :wk "5")
  "w 6" '((lambda () (interactive) (persp-frame-switch "6")) :wk "6")
  "w 7" '((lambda () (interactive) (persp-frame-switch "7")) :wk "7")
  "w 8" '((lambda () (interactive) (persp-frame-switch "8")) :wk "8")
  "w 9" '((lambda () (interactive) (persp-frame-switch "9")) :wk "9")
  "w w" '((lambda () (interactive) (persp-frame-switch "w")) :wk "w")
)
#+end_src

** anzu
#+begin_src emacs-lisp
(use-package anzu
  :ensure t
  :config
)

(use-package evil-anzu
  :ensure t
  :after (evil anzu)
  :config
  (setq anzu-search-threshold 0)
  (setq anzu-cons-mode-line-p nil)
  (global-anzu-mode 1))
#+end_src

** doom-modeline
#+begin_src emacs-lisp
(require 'doom-modeline)
(doom-modeline-mode 1)
#+end_src

** Perfect margin
#+begin_src emacs-lisp
(setq perfect-margin-visible-width 150)
#+end_src

** centered-cursor-mode
#+begin_src emacs-lisp
(use-package centered-cursor-mode
  :ensure t
  :config
  (global-centered-cursor-mode 0))
#+end_src


* KeyBinds
** file
#+begin_src emacs-lisp
(defun open-current-directory-in-dired ()
  "Open the current directory of the current file in Dired.
If a Dired buffer is already opened, navigate to that buffer."
  (interactive)
  (let ((dir (file-name-directory (or buffer-file-name default-directory))))
    (if (get-buffer-window (dired-noselect dir))
        (switch-to-buffer-other-window (dired-noselect dir))
      (dired-jump nil dir))))

(spc-leader
  "f" '(:ignore t :wk "file")
  "f c" '(copy-file :wk "copy file")
  "f o" '(open-current-directory-in-dired :wk "open in dired"))
#+end_src
** resize window
*** centered with dired
#+begin_src emacs-lisp
(defun centered ()
  "In dired, open current file in another window if exists, otherwise split window horizontally."
  (interactive)

  (split-window-right)
  (projectile-dired)
  (other-window 1)
  (enlarge-window-horizontally 70))

(spc-leader
  "c c" '(centered :wk "centered"))
#+end_src

** marks
#+begin_src emacs-lisp
(spc-leader
  "m" '(:ignore t :wk "marks")
  "m a" '(bm-toggle :wk "toggle mark")
  "m n" '(bm-next :wk "next mark")
  "m p" '(bm-previous :wk "prev mark")
  "m l" '(bm-show :wk "list marks")
  "m g" '(lambda () (interactive) (bm-show-goto-bookmark) (recenter) :wk "goto bookmark")
  "m k" '(bm-remove-all-all-buffers :wk "claer mark"))

(setq bm-marker 'bm-marker-left)
(setq bm-highlight-style nil)
(setq bm-highlight-style 'bm-highlight-only-fringe)
(custom-set-faces
 '(bm-face ((t (:background "yellow" :foreground "black" :weight bold)))))
#+end_src

** buffer & bookmarks & harpoon
#+begin_src emacs-lisp
(spc-leader
  ;; buffer
  "b" '(:ignore t :wk "buffer")
  "b i" '(ibuffer :wk "Switch ibuffer")
  "b b" '(projectile-switch-to-buffer :wk "Switch buffer")
  "b k" '(kill-this-buffer :wk "Kill this buffer")
  ;; bookmark
  "b s" '(bookmark-set :wk "set bookmarks")
  "b d" '(bookmark-delete :wk "delete bookmarks")
  "b l" '(list-bookmarks :wk "list bookmarks")
  "b j" '(bookmark-jump :wk "list jump")
  "b a" '(bookmark-set :wk "bookmarks set")

  "b r" '(revert-buffer :wk "Reload buffer"))

(spc-leader
  "TAB" '(:ignore t :wk "?")
  "TAB TAB" '(comment-line :wk "comment line")
)

(spc-leader
  "h" '(:ignore t :wk "help & harpoon")
  "h a" '(harpoon-add-file :wk "add to harpoon")
  "h c" '(harpoon-clear :wk "harboon clear")
  "h l" '(harpoon-toggle-quick-menu :wk "toggle menu")
  "h m" '(harpoon-quick-menu-hydra :wk "quick menu")

  "h 1" '(harpoon-go-to-1 :wk "1")
  "h 2" '(harpoon-go-to-2 :wk "2")
  "h 3" '(harpoon-go-to-3 :wk "3")
  "h 4" '(harpoon-go-to-4 :wk "4")
  "h 5" '(harpoon-go-to-5 :wk "5")

  "h d" '(:ignore t :wk "harpoon delete")
  "h d 1" '(harpoon-delete-1 :wk "1")
  "h d 2" '(harpoon-delete-2 :wk "2")
  "h d 3" '(harpoon-delete-3 :wk "3")
  "h d 4" '(harpoon-delete-4 :wk "4")
  "h d 5" '(harpoon-delete-5 :wk "5")

  "h r r" '((lambda () (interactive) (load-file "/home/fs/.emacs.d/init.el")) :wk "Reload emacs config"))
#+end_src

** ivy
#+begin_src emacs-lisp
(define-key ivy-minibuffer-map (kbd "C-j") 'ivy-next-line)
(define-key ivy-minibuffer-map (kbd "C-k") 'ivy-previous-line)
#+end_src

** project
#+begin_src emacs-lisp
(spc-leader
  "," '(find-file :wk "fzf find file")
  "SPC" '(counsel-fzf :wk "fzf find file")
  "." '(swiper :wk "swiper")
  "p" '(:ignore t :wk "projectile")
  "p a" '(projectile-add-known-project :wk "add project")
  "p s" '(projectile-switch-project :wk "switch project")
  "p g" '(counsel-rg :wk "project tile grep")

  "b i" '(projectile-ibuffer :wk "Switch buffer")
  "b n" '(projectile-next-project-buffer :wk "next project buffer")
  "b p" '(projectile-previous-project-buffer :wk "previous project buffer"))

#+end_src

** DIRED
*** default dired
#+begin_src emacs-lisp
(spc-leader
  "d" '(:ignore t :wk "dired")
  "d o" '(dired :wk "open dired")

  "d c" '(:ignore t :wk "copy/crete")
  "d c c" '(dired-do-copy :wk "copy")
  "d c d" '(dired-create-directory :wk "create dir")
  "d c f" '(dired-create-empty-file :wk "crete file")
  "d r" '(dired-do-rename :wk "renema & move")

  "d p" '(:ignore t :wk "permissions")
  "d p p" '(dired-do-chmod :wk "chmod")
  "d p o" '(dired-do-chown :wk "chown")

  "d h" '(dired-hide-details-mode :wk "hide/show deteils")
)
#+end_src

открыть файл в вертикально разделенном экране
#+begin_src emacs-lisp
(defun dired-open-split-horizontal ()
  "In dired, open current file in another window if exists, otherwise split window horizontally."
  (interactive)
  ;; Сохраняем путь текущего файла
  (let ((current-file (dired-get-file-for-visit)))
    ;; Проверяем, есть ли другое окно
    (if (one-window-p)
        (split-window-horizontally))
    ;; Переключаемся на другое окно
    (other-window 1)
    ;; Открываем файл или каталог, используя сохраненный путь
    (find-file current-file))
    (other-window 1))

(define-key dired-mode-map (kbd "s") 'dired-open-in-split-window)

(evil-define-key 'normal dired-mode-map (kbd "s") 'dired-open-split-horizontal)
(evil-define-key 'normal dired-mode-map (kbd "h") 'dired-up-directory)
#+end_src

*** dired subtree
#+begin_src emacs-lisp
(require 'dired-subtree)

(evil-define-key 'normal dired-mode-map (kbd "i") 'dired-subtree-insert)
(evil-define-key 'normal dired-mode-map (kbd "r") 'dired-subtree-remove)
(evil-define-key 'normal dired-mode-map (kbd "o") 'dired-open-file-in-new-tab)

#+end_src


** Sudo Edit & code
#+begin_src emacs-lisp
(require 'sudo-edit)
(spc-leader
  "c" '(:ignore t :wk "code")
  "c s" '(sudo-edit :wk "sudo-edit")
  "s" '(:ignore t :wk "split window")
  "s h" '(split-window-right :wk "split window horizontaly")
  "s v" '(split-window-below :wk "split window verticalty")
  "c p" '(perfect-margin-mode :wk "center")
)
#+end_src

** ORG mode
#+begin_src emacs-lisp
(spc-leader
  "o" '(:ignore t :wk "org mode")
  "o c" '(org-time-stamp :wk "calendar paste date")
  "o l" '(org-shiftmetaright :wk "move metric right")
  "o h" '(org-shiftmetaleft :wk "move metric left")
  "o t" '(org-tree-to-indirect-buffer :wk "tree to inderect buffer")
  "o o" '(org-cycle-global :wk "org cycle global")
  "o i" '(org-display-inline-images :wk "display image")
)
#+end_src

** Restclient
#+begin_src emacs-lisp
(spc-leader
  "r" '(:ignore t :wk "rest client")
  "r s" '(restclient-http-send-current-stay-in-window :wk "send http"))
#+end_src

** Magit
#+begin_src emacs-lisp

(require 'magit)
  (spc-leader
    "g" '(:ignore t :wk "git")
    "g g" '(magit :wk "magit")
    "g a" '(magit-stage-buffer-file :wk "add curent file")
    "g u" '(magit-unstage-buffer-file :wk "unstage current file")

    "g l" '(:ignore t :wk "log")
    "g l g" '(magit-log-current :wk "log")
    "g l l" '(magit-log-buffer-file :wk "log buffer")

    "g b" '(:ignore t :wk "branch")
    "g b b" '(magit-branch-checkout t :wk "branch")
    "g b p" '(magit-pull-branch t :wk "branch")

    "g d" '(:ignore t :wk "diff")
    "g d d" '(magit-diff-buffer-file :wk "log buffer")
    "g d g" '(magit-diff-unstaged :wk "diff unstaged buffer")
    "g d s" '(magit-diff-staged :wk "diff staged buffer")
  )

  ; (use-package magit-delta
  ;   :hook (magit-mode . magit-delta-mode))

  ; (defun my-magit-toggle-evil-visual-line-mode ()
  ;   "Включает или выключает evil-visual-line-mode в режиме magit-mode."
  ;   (interactive)
  ;    (evil-visual-line))

  ; (define-key magit-mode-map (kbd "C-c v") #'my-magit-toggle-evil-visual-line-mode)

#+end_src

* ORG settings
** Org Temo chet sheet
| Typing the below + TAB | Expands to ...                          |
|------------------------+-----------------------------------------|
| <a                     | '#+BEGIN_EXPORT ascii' … '#+END_EXPORT  |
| <c                     | '#+BEGIN_CENTER' … '#+END_CENTER'       |
| <C                     | '#+BEGIN_COMMENT' … '#+END_COMMENT'     |
| <e                     | '#+BEGIN_EXAMPLE' … '#+END_EXAMPLE'     |
| <E                     | '#+BEGIN_EXPORT' … '#+END_EXPORT'       |
| <h                     | '#+BEGIN_EXPORT html' … '#+END_EXPORT'  |
| <l                     | '#+BEGIN_EXPORT latex' … '#+END_EXPORT' |
| <q                     | '#+BEGIN_QUOTE' … '#+END_QUOTE'         |
| <s                     | '#+BEGIN_SRC' … '#+END_SRC'             |
| <v                     | '#+BEGIN_VERSE' … '#+END_VERSE'         |

** toc
#+begin_src emacs-lisp
(require 'toc-org nil t)
   (add-hook 'org-mode-hook 'toc-org-enable)
   (add-hook 'org-mode-hook 'org-indent-mode)
   ;; (setq toc-org-max-depth 2)

(require 'org-bullets)
  (add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))

(require 'org-bullets)
  (which-key-mode 1)

    (setq which-key-min-display-lines 10)
    (setq which-key-side-window-location 'bottom
    which-key-sort-order #'which-key-key-order-alpha
    which-key-separator " → " )

(require 'org-tempo)

;;(electric-indent-mode -1)
(setq org-edit-src-content-indentation 0)
#+end_src

* PROGRAMMING
** python
#+begin_src emacs-lisp
(add-hook 'python-mode-hook
          (lambda ()
            (setq indent-tabs-mode nil)
            (setq tab-width 4)
            (setq python-indent-offset 4)
            (elpy-enable)
            (elpy-mode)
            (flymake-mode -1) ;; turn off flymake
            (remove-hook 'elpy-modules 'elpy-module-flymake)
            (add-hook 'elpy-mode-hook (lambda () (highlight-indentation-mode -1)))))
#+end_src

** tree-sitter
#+begin_src emacs-lisp
(add-hook 'python-mode-hook #'tree-sitter-mode +1)
(add-hook 'python-mode-hook #'tree-sitter-hl-mode +1)
#+end_src


* decorations
** colorthemes
#+begin_src emacs-lisp
(custom-set-faces
 '(line-number ((t (:foreground "white"))))               ; Цвет отображения номера строки
 '(line-number-current-line ((t (:foreground "white"))))) ; Цвет текущей строки

; (load-theme 'doom-nord t)
; (load-theme 'doom-nord-aurora t)
; (load-theme 'doom-nord-light t)
; (load-theme 'doom-gruvbox t)
; (load-theme 'doom-ayu-dark t)
; (load-theme 'doom-dracula t)
; (load-theme 'doom-tomorrow-night t)
; (load-theme 'doom-monokai-pro t)
(load-theme 'doom-spacegrey t)

; (add-to-list 'custom-theme-load-path "~/.emacs.d/everforest-theme")
; (load-theme 'everforest-hard-dark t)

(set-frame-parameter nil 'alpha-background 92)
#+End_src

** diff
#+begin_src emacs-lisp
(global-diff-hl-mode +1)
#+end_src

** fonts
#+begin_src emacs-lisp
;; Makes commented text and keywords italics.
;; This is working in emacsclient but not emacs.
;; Your font must have an italic face available.
(set-face-attribute 'font-lock-comment-face nil
  :slant 'italic)
(set-face-attribute 'font-lock-keyword-face nil
  :slant 'italic)

(defun set-small-font ()
  (interactive)
  (set-face-attribute 'default nil
                      :font "JetBrainsMonoNL Nerd Font"
                      :height 110
                      :weight 'medium)
  (set-face-attribute 'variable-pitch nil
                      :font "JetBrainsMonoNL Nerd Font"
                      :height 120
                      :weight 'medium)
  (set-face-attribute 'fixed-pitch nil
                      :font "JetBrainsMonoNL Nerd Font"
                      :height 110
                      :weight 'medium))

(defun set-big-font ()
  (interactive)
  (set-face-attribute 'default nil
                      :font "JetBrainsMonoNL Nerd Font"
                      :height 150
                      :weight 'medium)
  (set-face-attribute 'variable-pitch nil
                      :font "JetBrainsMonoNL Nerd Font"
                      :height 160
                      :weight 'medium)
  (set-face-attribute 'fixed-pitch nil
                      :font "JetBrainsMonoNL Nerd Font"
                      :height 150
                      :weight 'medium))

(spc-leader
  "c f" '(:ignore t :wk "font")
  "c f s" '(set-small-font :wk "small font")
  "c f b" '(set-big-font :wk "big font"))
;; This sets the default font on all graphical frames created after restarting Emacs.
;; Does the same thing as 'set-face-attribute default' above, but emacsclient fonts
;; are not right unless I also add this method of setting the default font.
(add-to-list 'default-frame-alist '(font . "JetBrainsMonoNL Nerd Font"))

;; Uncomment the following line if line spacing needs adjusting.
(setq-default line-spacing 0.12)

#+end_src


** icons
#+begin_src emacs-lisp
(require 'all-the-icons)
(require 'all-the-icons-dired)
(add-hook 'dired-mode-hook 'all-the-icons-dired-mode)
#+end_src

** rainbow mode
This minor mode sets background color to strings that match color
names, e.g. #0000ff is displayed in white with a blue background.
#+begin_src emacs-lisp
(require 'rainbow-mode)
(add-hook 'org-mode-hook 'rainbow-mode)
(add-hook 'prog-mode-hook 'rainbow-mode)
#+end_src

** Nyan-mode
#+begin_src emacs-lisp
(require 'nyan-mode)
(nyan-mode +1)
(setq nyan-toggle-wavy-trail t)
(setq nyan-start-animation t)
#+end_src

** evil-snipe
#+begin_src emacs-lisp
(require 'evil-snipe)
(evil-snipe-mode +1)
(evil-snipe-override-mode +1)
(setq evil-snipe-smart-case +1)
#+end_src

** org chckbox hl
#+begin_src emacs-lisp
(defface org-checkbox-done-text
     '((t (:inherit org-done)))
     "Face for the text part of a checked org-mode checkbox.")

 (font-lock-add-keywords
  'org-mode
  `(("^[ \t]*\\(?:[-+*]\\|[0-9]+[).]\\)[ \t]+\\(\\(?:\\[@\\(?:start:\\)?[0-9]+\\][ \t]*\\)?\\[\\(?:X\\|\\([0-9]+\\)/\\2\\)\\][^\n]*\n\\)" 1 'org-checkbox-done-text prepend))
  'append)
#+end_src

** hl-todo
#+begin_src emacs-lisp
(require 'hl-todo)
(global-hl-todo-mode +1)
#+end_src


home/.emacs.d/config.el
(defvar my-packages
  '(dracula-theme
    evil
    evil-collection
    general
    key-chord
    toc-org
    org-bullets
    projectile
   )
    "A list of packages to ensure are installed at launch.")

  (defun my-packages-installed-p ()
    (let ((all-installed t))
      (dolist (p my-packages all-installed)
        (unless (package-installed-p p)
          (setq all-installed nil)))))

  (unless (my-packages-installed-p)
    (message "%s" "Refreshing package database...")
    (package-refresh-contents)
    (dolist (p my-packages)
      (unless (package-installed-p p)
        (package-install p))))

(menu-bar-mode -1)
(tool-bar-mode -1)
(scroll-bar-mode -1)

(global-visual-line-mode t)
(setq-default display-line-numbers 'relative)

(require 'package)
  (add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
  (add-to-list 'package-archives
	   '("melpa-stable" . "https://stable.melpa.org/packages/") t)
(package-initialize)

;; evil, evil-collection

(setq evil-want-integration t) ;; This is optional since it's already set to t by default.
(setq evil-want-keybinding nil)
(setq evil-vsplit-window-right t)
(setq evil-split-window-below t)

(setq evil-want-C-u-scroll t)

(require 'evil)
(evil-mode 1)

(setq evil-collection-mode-list '(dashboard dired ibuffer))
(evil-collection-init)

(require 'general)
(general-evil-setup)

(general-create-definer spc-leader
  :states '(normal insert visual emacs)
  :keymaps 'override
  :prefix "SPC" ;; set leader
  :global-prefix "M-SPC") ;; access leader in insert mode

(spc-leader
  "b" '(:ignore t :wk "buffer")
  "b b" '(switch-to-buffer :wk "Switch buffer")
  "b i" '(ibuffer :wk "Switch buffer")
  "b k" '(kill-this-buffer :wk "Kill this buffer")
  "b n" 'next-buffer
  "b p" '(previous-buffer :wk "Previous buffer")
  "b r" '(revert-buffer :wk "Reload buffer"))
(spc-leader
  "h" '(:ignore t:wk "Help")
  "h r r" '((lambda () (interactive) (load-file "/home/fs/.emacs.d/init.el")) :wk "Reload emacs config"))

(require 'key-chord)
(key-chord-mode 1)
(require 'evil)

(setq key-chord-two-keys-delay 0.03) ; 30ms
  (key-chord-define evil-insert-state-map "jk" 'evil-normal-state)

(setq key-chord-two-keys-delay 0.03) ; 30ms
  (key-chord-define evil-insert-state-map "kj" 'evil-normal-state)

(require 'toc-org nil t)
   (add-hook 'org-mode-hook 'toc-org-enable)
   (add-hook 'org-mode-hook 'org-indent-mode)

(require 'org-bullets)
  (add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))

(require 'org-bullets)
  (which-key-mode 1)
    (setq which-key-side-window-location 'bottom
    which-key-sort-order #'which-key-key-order-alpha
    which-key-sort-uppercase-first nil
    which-key-add-column-padding 1
    which-key-max-display-columns nil
    which-key-min-display-lines 6
    which-key-side-window-slot -10
    which-key-side-window-max-height 0.25
    which-key-idle-delay 0.8
    which-key-max-description-length 25
    which-key-allow-imprecise-window-fit t
    which-key-separator " → " )

(electric-indent-mode -1)

(require 'org-tempo)

(load-theme 'dracula t)

(set-face-attribute 'default nil
  :font "JetBrainsMonoNL Nerd Font"
  :height 150 ;; 110
  :weight 'medium)
(set-face-attribute 'variable-pitch nil
  :font "JetBrainsMonoNL Nerd Font"
  :height 160 ;; 120
  :weight 'medium)
(set-face-attribute 'fixed-pitch nil
  :font "JetBrainsMonoNL Nerd Font"
  :height 150 ;; 110
  :weight 'medium)
;; Makes commented text and keywords italics.
;; This is working in emacsclient but not emacs.
;; Your font must have an italic face available.
(set-face-attribute 'font-lock-comment-face nil
  :slant 'italic)
(set-face-attribute 'font-lock-keyword-face nil
  :slant 'italic)

;; This sets the default font on all graphical frames created after restarting Emacs.
;; Does the same thing as 'set-face-attribute default' above, but emacsclient fonts
;; are not right unless I also add this method of setting the default font.
(add-to-list 'default-frame-alist '(font . "JetBrainsMonoNL Nerd Font"))

;; Uncomment the following line if line spacing needs adjusting.
(setq-default line-spacing 0.12)
